File: Chapter04/C#/CommandStack.cs
using System;
using System.Collections.Generic;

namespace EverydayDataStructures
{
	public class CommandStack
	{
		public Stack<Command> _commandStack { get; private set; }
		int _capacity;

		public CommandStack(int commandCapacity)
		{
			this._commandStack = new Stack<Command>(commandCapacity);
			this._capacity = commandCapacity;
		}

		public bool IsFull()
		{
			return this._commandStack.Count >= this._capacity;
		}

		public bool IsEmpty()
		{
			return this._commandStack.Count == 0;
		}

		public bool PerformCommand(Command command)
		{
			if (!this.IsFull())
			{
				this._commandStack.Push(command);
				return true;
			}
			return false;
		}

		public bool PerformCommands(List<Command> commands)
		{
			bool inserted = true;
			foreach (Command c in commands)
			{
				inserted = this.PerformCommand(c);
			}
			return inserted;
		}

		public Command UndoCommand()
		{
			return this._commandStack.Pop();
		}

		public void Reset()
		{
			this._commandStack.Clear();
		}

		public int TotalCommands()
		{
			return this._commandStack.Count;
		}
	}
}

File: Chapter04/Swift/CommandStack.swift
//
//  CommandStack.swift
//  EverydayDataStructures-Swift
//
//  Created by William Smith on 8/28/16.
//  Copyright © 2016 Websmiths, LLC. All rights reserved.
//

import Foundation

open class CommandStack
{
    open fileprivate(set) var _commandStack: Array = [Command]()
    open fileprivate(set) var _capacity: Int;

    public init (commandCapacity: Int)
    {
        _capacity = commandCapacity
    }

    open func IsFull() -> Bool
    {
        return _commandStack.count >= _capacity
    }

    open func IsEmpty() -> Bool
    {
        return _commandStack.count == 0
    }

    open func PerformCommand(_command: Command) -> Bool
    {
        if (!IsFull())
        {
            _commandStack.append(command)
            return true
        }
        return false
    }

    open func PerformCommands(_commands: [Command]) -> Bool
    {
        var inserted: Bool = true;
        for c in commands
        {
            inserted = PerformCommand(c)
        }
        return inserted
    }

    open func UndoCommand() -> Command
    {
        return _commandStack.popLast()!
    }

    open func Reset()
    {
        _commandStack.removeAll()
    }

    open func TotalCommands() -> Int
    {
        return _commandStack.count
    }
}
